src/libraries/System.Runtime.Numerics/src/System/Numerics/BigIntegerCalculator.SquMul.cs

```diff
@@ -173,161 +250,306 @@ public static void Multiply(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right, S
             // which are smaller in DEBUG mode for testing purpose.
 
             if (right.Length < MultiplyKaratsubaThreshold)
-            {
                 Naive(left, right, bits);
-                return;
+            else if ((left.Length + 1) >> 1 is int n && right.Length <= n)
+                RightSmall(left, right, bits, n);
+            else if (right.Length < MultiplyToom3Threshold)
+                Karatsuba(left, right, bits, n);
+            else
+                Toom3(left, right, bits);
+
+            static void Toom3(ReadOnlySpan<uint> left, ReadOnlySpan<uint> right, Span<uint> bits)
+            {
+                Debug.Assert(left.Length >= 3);
+                Debug.Assert(left.Length >= right.Length);
+                Debug.Assert(bits.Length >= left.Length + right.Length);
+                Debug.Assert(bits.Trim(0u).IsEmpty);
+
+                // Based on the Toom-Cook multiplication.
+                // Replace m in Wikipedia with left and n in Wikipedia with right.
+                // https://en.wikipedia.org/wiki/Toom-Cook_multiplication
+
+                int n = (left.Length + 2) / 3;
+
+                Debug.Assert(right.Length > n);
+                if (((uint)right.Length << 1) <= (uint)n)
+                {
+                    Toom25(left, right, bits, n);
+                    return;
+                }
+
+                int pLength = n + 1;
+                int pAndQAllLength = pLength * 6;
+                uint[]? pAndQAllFromPool = null;
+                Span<uint> pAndQAll = (
+                    (uint)pAndQAllLength <= StackAllocThreshold
+                    ? stackalloc uint[StackAllocThreshold]
+                    : pAndQAllFromPool = ArrayPool<uint>.Shared.Rent(pAndQAllLength)).Slice(0, pAndQAllLength);
```

Comment:
Given Toom3 is only selected with a larger threshold, consider to always use pooled array to make buffer management a bit simpler.